File: python/pointcloud.py
"""
PointCloud class.
"""

import numpy as np
from scipy import spatial


class PointCloud:
    """Class for working with point cloud data."""

    def __init__(self, x: np.array, y: np.array, z: np.array):
        self.x = x
        self.y = y
        self.z = z

        self.nx = None
        self.ny = None
        self.nz = None
        self.planarity = None

        self.no_points = len(x)
        self.sel = np.arange(0, len(x), 1)

    def select_in_range(self, X: np.array, max_range: float):
        """Select points within range of points X."""

        assert np.shape(X)[1] == 3, "X must have 3 columns!"

        kdtree = spatial.cKDTree(X)

        distances, _ = kdtree.query(
            self.X_sel, k=1, p=2, distance_upper_bound=max_range, workers=-1
        )

        keep = np.isfinite(distances)  # distances > max_range are infinite

        self.sel = self.sel[keep]

    def select_n_points(self, n: int):
        """Select n points equidistantly."""

        if self.no_selected_points > n:
            idx = np.round(np.linspace(0, self.no_selected_points - 1, n)).astype(int)
            self.sel = self.sel[idx]

    def estimate_normals(self, neighbors: int):
        """Estimate normal vectors for selected points from its neighborhood."""

        self.nx = np.full(self.no_points, np.nan)
        self.ny = np.full(self.no_points, np.nan)
        self.nz = np.full(self.no_points, np.nan)
        self.planarity = np.full(self.no_points, np.nan)

        kdtree = spatial.cKDTree(self.X)
        _, idxNN_all_qp = kdtree.query(self.X_sel, k=neighbors, p=2, workers=-1)

        for (i, idxNN) in enumerate(idxNN_all_qp):
            selected_points = np.column_stack(
                (self.x[idxNN], self.y[idxNN], self.z[idxNN])
            )
            C = np.cov(selected_points.T, bias=False)
            eig_vals, eig_vecs = np.linalg.eig(C)
            idx_sort = eig_vals.argsort()[::-1]  # sort from large to small
            eig_vals = eig_vals[idx_sort]
            eig_vecs = eig_vecs[:, idx_sort]
            self.nx[self.sel[i]] = eig_vecs[0, 2]
            self.ny[self.sel[i]] = eig_vecs[1, 2]
            self.nz[self.sel[i]] = eig_vecs[2, 2]
            self.planarity[self.sel[i]] = (eig_vals[1] - eig_vals[2]) / eig_vals[0]

    def transform(self, H: np.array):
        """Transform point cloud by given homogeneous transformation matrix H."""

        XInH = PointCloud.euler_coord_to_homogeneous_coord(self.X)
        XOutH = np.transpose(H @ XInH.T)
        XOut = PointCloud.homogeneous_coord_to_euler_coord(XOutH)

        self.x = XOut[:, 0]
        self.y = XOut[:, 1]
        self.z = XOut[:, 2]

    @property
    def x_sel(self) -> np.array:
        """Returns x coordinates of selected points."""
        return self.x[self.sel]

    @property
    def y_sel(self) -> np.array:
        """Returns y coordinates of selected points."""
        return self.y[self.sel]

    @property
    def z_sel(self) -> np.array:
        """Returns z coordinates of selected points."""
        return self.z[self.sel]

    @property
    def X(self) -> np.array:
        """Returns n-by-3 matrix of all points."""
        return np.column_stack((self.x, self.y, self.z))

    @property
    def X_sel(self) -> np.array:
        """Returns n-by-3 matrix of selected points."""
        return np.column_stack((self.x[self.sel], self.y[self.sel], self.z[self.sel]))

    @property
    def no_selected_points(self) -> int:
        """Returns number of selected points."""
        return len(self.sel)

    @staticmethod
    def euler_coord_to_homogeneous_coord(XE: np.array):
        """Convert Euler coordinates to homogeneous coordinates."""

        no_points = np.shape(XE)[0]
        XH = np.column_stack((XE, np.ones(no_points)))

        return XH

    @staticmethod
    def homogeneous_coord_to_euler_coord(XH: np.array):
        """Convert homogeneous coordinates to Euler coordinates."""

        XE = np.column_stack(
            (XH[:, 0] / XH[:, 3], XH[:, 1] / XH[:, 3], XH[:, 2] / XH[:, 3])
        )

        return XE

File: matlab/pointcloud.m
classdef pointcloud < handle

    properties

        x
        y
        z

        nx
        ny
        nz
        planarity

        noPoints
        sel

    end

    methods

        function obj = pointcloud(x, y, z)

            obj.x = x;
            obj.y = y;
            obj.z = z;

            obj.noPoints = numel(x);
            obj.sel = transpose(1:obj.noPoints);

        end
        
        function selectInRange(obj, X, range)
            
            [~, distance] = knnsearch(X, [obj.x(obj.sel) obj.y(obj.sel) obj.z(obj.sel)], ...
                'BucketSize', 1000);
            keep = distance <= range;
            obj.sel = obj.sel(keep);
            
        end

        function selectNPoints(obj, n)

            noSelectedPoints = numel(obj.sel);
            
            if noSelectedPoints > n
                idx = transpose(round(linspace(1, noSelectedPoints, n)));
                obj.sel = obj.sel(idx);
            end

        end

        function estimateNormals(obj, neighbors)

            obj.nx = NaN(obj.noPoints,1);
            obj.ny = NaN(obj.noPoints,1);
            obj.nz = NaN(obj.noPoints,1);
            obj.planarity = NaN(obj.noPoints,1);

            idxNN = knnsearch(...
                [obj.x obj.y obj.z], ...
                [obj.x(obj.sel) obj.y(obj.sel) obj.z(obj.sel)], ...
                'k', neighbors);

            for i = 1:numel(obj.sel)
                C = cov([obj.x(idxNN(i,:)) obj.y(idxNN(i,:)) obj.z(idxNN(i,:))]);
                [P, latent] = pcacov(C);
                obj.nx(obj.sel(i)) = P(1,3);
                obj.ny(obj.sel(i)) = P(2,3);
                obj.nz(obj.sel(i)) = P(3,3);
                obj.planarity(obj.sel(i)) = (latent(2)-latent(3))/latent(1);
            end

        end

        function transform(obj, H)

            XInH = obj.eulerCoordToHomogeneousCoord([obj.x obj.y obj.z]);
            XOutH = transpose(H*XInH');
            XOut = obj.homogeneousCoordToEulerCoord(XOutH);

            obj.x = XOut(:,1);
            obj.y = XOut(:,2);
            obj.z = XOut(:,3);

        end

    end

    methods(Static)

        function XH = eulerCoordToHomogeneousCoord(XE)

            noPoints = size(XE, 1);
            XH = [XE ones(noPoints,1)];

        end

        function XE = homogeneousCoordToEulerCoord(XH)

            XE = XH(:,[1 2 3])./XH(:,4);

        end

    end

end
